crates/re_chunk/src/chunk.rs

```diff
@@ -0,0 +1,478 @@
+use std::collections::BTreeMap;
+
+use arrow2::array::Array as ArrowArray;
+
+use re_log_types::{EntityPath, ResolvedTimeRange, RowId, TimeInt, TimePoint, Timeline};
+use re_types_core::{ComponentName, SerializationError};
+
+// ---
+
+/// Errors that can occur when creating/manipulating a [`Chunk`]s.
+#[derive(thiserror::Error, Debug)]
+pub enum ChunkError {
+    #[error("Detected malformed Chunk: {reason}")]
+    Malformed { reason: String },
+
+    #[error(transparent)]
+    Serialization(#[from] SerializationError),
+
+    #[error("Chunks cannot be empty")]
+    Empty,
+}
+
+pub type ChunkResult<T> = Result<T, ChunkError>;
+
+// ---
+
+/// Unique identifier for a [`Chunk`], using a [`re_tuid::Tuid`].
+pub type ChunkId = re_tuid::Tuid;
+
+/// Dense arrow-based storage of N rows of multi-component multi-temporal data for a specific entity.
+///
+/// This is our core datastructure for logging, storing, querying and transporting data around.
+///
+/// The chunk as a whole is always ascendingly sorted by [`RowId`] before it gets manipulated in any way.
+/// Its time columns might or might not be ascendingly sorted, depending on how the data was logged.
+///
+/// This is the in-memory representation of a chunk, optimized for efficient manipulation of the
+/// data within.
+#[derive(Debug, Clone)]
+pub struct Chunk {
+    pub(crate) id: ChunkId,
```

Comment:
I'm exploring the possibility of always making sure that the ID of a chunk is the same as the ID of its first row (in sorted order).

That would be way more useful than a random ID generated post-micro-batching, and would give way more meaning to sorting chunks based on their IDs.